subsys/usb/usb_descriptor.c

```diff
@@ -290,15 +292,26 @@ static struct usb_cfg_data *usb_get_cfg_data(struct usb_if_descriptor *iface)
 }
 
 /*
- * Default USB SN string descriptor is CONFIG_USB_DEVICE_SN, but sometimes
- * we want use another string as SN descriptor such as the chip's unique
- * ID. So user can implement this function return a string thats will be
- * replaced the default SN.
- * Please note that the new SN descriptor you changed must has same length
- * as CONFIG_USB_DEVICE_SN.
+ * Default USB Serial Number string descriptor will be derived from
+ * Hardware Information Driver (HWINFO). User can implement own variant
+ * of this function. Please note that the length of the new Serian Number
+ * descriptor may not exceed the length of the CONFIG_USB_DEVICE_SN.
  */
 __weak u8_t *usb_update_sn_string_descriptor(void)
 {
+	u8_t hwid[sizeof(CONFIG_USB_DEVICE_SN) / 2];
+
+	memset(hwid, 0, sizeof(hwid));
+	if (hwinfo_get_device_id(hwid, sizeof(hwid)) > 0) {
+		LOG_HEXDUMP_DBG(hwid, sizeof(hwid), "Serial Number");
+
+		snprintf(string_descr.utf16le_sn.bString,
```

Comment:
Thanks for you talk which pointed me to this bug report! I fully support adding code to automatically fill the serial number of USB devices.

`hwinfo_get_device_id` might return a smaller ID than `sizeof(hwid)`. In that case the the USB serial number will be smaller, not sure that's a real issue and if any of the Zephyr drivers actually return a so small ID. OTOH, I wonder if some drivers might report a longer ID implemented on the silicon in a more or less incremental way, so that two SoCs produced in the same batch might have an ID with an identical start.

In that regard, I wonder if it might be better to pass the full ID through a hash function like CRC32 (as we already have such a hash function in Zephyr) and use that instead. That should provide an invariant ID, but still independent on how the manufacturers create the ID.

It might actually be a good idea to implement such a function directly in the hwinfo driver. Some functions returning a 4, 16 or 32 bytes ID.

Reply:
@aurel32 I agree, but I would like to keep it simpler here. @alexanderwachter FYI :arrow_up:

Reply:
@aurel32 I had precisely the same Idea, regarding the 16-, 32-bit values and CRCs. However, I decided to leave that up to the user. The "problem" SoCs are the STM32 series. They use a combination of batch number and the production date and things like that. Only in the last few bits, there are the X and Y coordinates on the wavers.